modules/src/core/ics04_channel/packet.rs

```diff
@@ -188,19 +198,46 @@ impl core::fmt::Display for Packet {
     }
 }
 
+/// Parse a string into a timeout height expected to be stored in
+/// `Packet.timeout_height`. We need to parse the timeout height differently
+/// because of a quirk introduced in ibc-go. See comment in
+/// `TryFrom<RawPacket> for Packet`.
+pub fn parse_timeout_height(s: &str) -> Result<TimeoutHeight, Error> {
+    match s.parse::<Height>() {
+        Ok(height) => Ok(TimeoutHeight::from(height)),
+        Err(e) => match e.into_detail() {
+            HeightErrorDetail::ZeroHeight(_) => Ok(TimeoutHeight::no_timeout()),
+            _ => Err(Error::invalid_timeout_height()),
+        },
+    }
+}
+
 impl TryFrom<RawPacket> for Packet {
     type Error = Error;
 
     fn try_from(raw_pkt: RawPacket) -> Result<Self, Self::Error> {
         if Sequence::from(raw_pkt.sequence).is_zero() {
             return Err(Error::zero_packet_sequence());
         }
-        let packet_timeout_height: Height = raw_pkt
+
+        // `RawPacket.timeout_height` is treated differently because
+        //
+        // `RawHeight.timeout_height == {revision_number: 0, revision_height = 0}`
+        //
+        // is legal and meaningful, even though the Tendermint spec rejects this
+        // height as invalid. Thus, it must be parsed specially, where this
+        // special case means "no timeout".
+        //
+        // Note: it is important for `revision_number` to also be `0`, otherwise
+        // packet commitment proofs will be incorrect (see proof construction in
+        // `ChannelReader::packet_commitment()`). Note also that ibc-go conforms
+        // to this.
+        let packet_timeout_height: TimeoutHeight = raw_pkt
             .timeout_height
-            .ok_or_else(Error::missing_height)?
-            .into();
+            .try_into()
+            .map_err(|_| Error::invalid_timeout_height())?;
 
-        if packet_timeout_height.is_zero() && raw_pkt.timeout_timestamp == 0 {
+        if !packet_timeout_height.has_timeout() && raw_pkt.timeout_timestamp == 0 {
```

Comment:
Unrelated to this PR, but I think the spec allows a packet without a timeout height and timestamp, so this is a bug and we shouldn't have this check here.